### PR TITLE
Minimized False Positives in Format String and Buffer Overflow

### PR DESCRIPTION
**Pull request by msrader**
I added code to dump out of Format String and Buffer Overflow if it detects a server error before running the scanner.  I also upgraded the exception handling.  Two commits are used.  One for each scanner.

### CODE REVIEW

**Comment by msrader** [comment]
This was written to prevent a bad server request from causing a false positive.  This was the result of some emails I interchanged with one of the primary developers.  "I also wanted to ask you if you have some time to do some changes to the scanners IntegerOverflow, BufferOverflow and FormatString. Currentlythey lead to FPs if the original request already returns 500 status code"

**Comment by kingthorin** [comment]
Yes 500 is a subset of the isServerError check, if it addresses the feedback you already got then disregard my comment.

File: src/org/zaproxy/zap/extension/ascanrules/BufferOverflow.java [hunk 1]
```diff
@@ -62,7 +66,12 @@ public String getName() {
 	public String[] getDependency() {
 		return null;
 	}
-
+	
+	@Override
+	public boolean targets(TechSet technologies) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/BufferOverflow.java]: This should be mentioned in the commit message (or done in another commit).
I guess this should be mentioned in the help pages too.

File: src/org/zaproxy/zap/extension/ascanrules/BufferOverflow.java [hunk 2]
```diff
@@ -135,7 +148,7 @@ public void scan(HttpMessage msg, String param, String value) {
 				return;	
 
 			
-		} catch (Exception e) {
+		} catch (IOException e) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/BufferOverflow.java]: This change should be mentioned in the commit message (like it was done in the PR description).

File: src/org/zaproxy/zap/extension/ascanrules/FormatString.java [hunk 3]
```diff
@@ -251,7 +258,7 @@ public void scan(HttpMessage msg, String param, String value) {
 
 
 			
-		} catch (Exception e) {
+		} catch (IOException e) {
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/FormatString.java]: This change should be mentioned in the commit message (like it was done in the PR description).

File: src/org/zaproxy/zap/extension/ascanrules/ZapAddOn.xml [hunk 4]
```diff
@@ -1,40 +1,42 @@
-<zapaddon>
```
Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/ZapAddOn.xml]: There were some line endings changes in the first commit.

**Comment by thc202** [comment]
@kingthorin the check for 500 status code should be enough, since that's the main condition to raise the alert.

File: src/org/zaproxy/zap/extension/ascanrules/ZapAddOn.xml (hunk 4 above)
Comment (msrader) [inline src/org/zaproxy/zap/extension/ascanrules/ZapAddOn.xml]: Im not sure what you mean by this.  Did CRLF changes get added?

Comment (thc202) [inline src/org/zaproxy/zap/extension/ascanrules/ZapAddOn.xml]: Yes, the line endings were changed from LF to CRLF.

**Comment by thc202** [comment]
Thanks!

**Pull request merged**